### PR TITLE
feat: add payment method service v2

### PR DESCRIPTION
**Pull request by ligertsicat**
This PR add a Payment Method Service V2 client based on this stoplight
https://xendit.stoplight.io/docs/unified-payment-api/81a1e463f9dc2-create-a-payment-method

### CODE REVIEW

File: payment_method.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	Actions            PaymentMethodActions             `json:"actions"`
+	Description        *string                          `json:"description"`
+	CreatedAt          time.Time                        `json:"created"`
+	UpdatedAt          time.Time                        `json:"updated"`
+	Metadata           *Metadata                        `json:"metadata"`
+	BillingInformation *BillingInformation              `json:"billing_information"`
+	FailureCode        *string                          `json:"failure_code"`
+
+	Ewallet            *EwalletMethod            `json:"ewallet"`
+	DirectBankTransfer *DirectBankTransferMethod `json:"direct_bank_transfer"`
```
Comment (arlolanuzo) [inline payment_method.go:118]: can remove since this is not live yet

File: payment_method.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	DirectDebit        *DirectDebitMethod        `json:"direct_debit"`
+	Card               *card.Method              `json:"card"`
+	Cryptocurrency     *cryptocurrency.Method    `json:"cryptocurrency,omitempty"`
```
Comment (arlolanuzo) [inline payment_method.go:121]: can remove also since this is not live yet

File: pmsv2/virtualaccount/params.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	BCA               ChannelCode = "BCA"
+	BJB               ChannelCode = "BJB"
+	BNI               ChannelCode = "BNI"
+	BRI               ChannelCode = "BRI"
+	MANDIRI           ChannelCode = "MANDIRI"
+	PERMATA           ChannelCode = "PERMATA"
+	BSI               ChannelCode = "BSI"
+	CIMB              ChannelCode = "CIMB"
+	SAHABAT_SAMPOERNA ChannelCode = "SAHABAT_SAMPOERNA"
+	ARTAJASA          ChannelCode = "ARTAJASA"
```
Comment (arlolanuzo) [inline pmsv2/virtualaccount/params.go:21]: can remove since integration is canceled for ARTAJASA

File: pmsv2/virtualaccount/params.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+type CreateMethod struct {
+	Amount            float64               `json:"amount"`
+	MinimumAmount     float64               `json:"min_amount,omitempty"`
+	MaximumAmount     float64               `json:"max_amount,omitempty"`
+	Currency          constant.CurrencyEnum `json:"currency"`
+	ChannelCode       ChannelCode           `json:"channel_code"`
+	ChannelProperties ChannelProperties     `json:"channel_properties"`
+	VirtualAccountID  string                `json:"virtual_account_id"`
```
Comment (arlolanuzo) [inline pmsv2/virtualaccount/params.go:41]: can remove, since this is an internal field

File: payment_method.go (hunk 1 above)
Comment (ligertsicat) [reply payment_method.go:118]: removed direct bank transfer

File: payment_method.go (hunk 2 above)
Comment (ligertsicat) [reply payment_method.go:121]: removed cryptocurrency (and related folders)

File: pmsv2/virtualaccount/params.go (hunk 3 above)
Comment (ligertsicat) [reply pmsv2/virtualaccount/params.go]: removed

File: pmsv2/virtualaccount/params.go (hunk 4 above)
Comment (ligertsicat) [reply pmsv2/virtualaccount/params.go]: removed. Though, this is also available in the CreateMethod of PMSv2. Is that fine?
image

Comment (arlolanuzo) [reply pmsv2/virtualaccount/params.go]: hmm that's a bug then, ideally it should only be available in internal GET response.

File: pmsv2/client.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+func (c *Client) CreatePaymentMethodWithContext(ctx context.Context, data *CreatePaymentMethodParams) (*xendit.PaymentMethodResponse, *xendit.Error) {
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.XCallbackURL != "" {
+		header.Add("x-callback-url", data.XCallbackURL)
+	}
+	if data.ServiceID != nil {
+		header.Add("service-id", *data.ServiceID)
+	}
```
Comment (arlolanuzo) [inline pmsv2/client.go:32]: I think these are for internal-use only 🤔 let's remove also

File: pmsv2/client.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
+
+	if data.ServiceID != nil {
+		header.Add("service-id", *data.ServiceID)
+	}
```
Comment (arlolanuzo) [inline pmsv2/client.go:95]: ditto

File: pmsv2/constant/params.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	CurrencyVND CurrencyEnum = "VND"
+	CurrencyTHB CurrencyEnum = "THB"
+)
+
+type PaymentMethodStatusEnum string
+
+const (
+	Active         PaymentMethodStatusEnum = "ACTIVE"
+	Expired        PaymentMethodStatusEnum = "EXPIRED"
+	Expiring       PaymentMethodStatusEnum = "EXPIRING" //internal use only
```
Comment (arlolanuzo) [inline pmsv2/constant/params.go:37]: ditto

File: pmsv2/directdebit/params.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
+package directdebit
+
+type ChannelCode string
+type ChannelProperties map[string]interface{}
+type Type string
+
+const (
```
Comment (arlolanuzo) [inline pmsv2/directdebit/params.go:7]: let's also remove the channel codes that are not yet supported by PMSv2 - [reference here](https://xendit.atlassian.net/wiki/spaces/CHARGE/pages/2654535704/Supported+Channels)

Comment (arlolanuzo) [reply pmsv2/directdebit/params.go:7]: can we also do this to other payment channels?

File: pmsv2/overthecounter/params.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	PaymentCode  string     `json:"payment_code"`
+	ExpiresAt    *time.Time `json:"expires_at,omitempty"`
+}
+
+type CreateMethod struct {
+	Amount            float64               `json:"amount"`
+	Currency          constant.CurrencyEnum `json:"currency"`
+	ChannelCode       ChannelCode           `json:"channel_code"`
+	ChannelProperties ChannelProperties     `json:"channel_properties"`
+	PaymentCodeID     string                `json:"payment_code_id"`
```
Comment (arlolanuzo) [inline pmsv2/overthecounter/params.go:44]: for internal-use also, no need to expose

File: pmsv2/params.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+	Code string `json:"auth_code"`
+
+	MerchantID string `json:"-"`
+}
+
+type ExpireRequest struct {
+	ID                     string  `json:"-"`
+	UnlinkSuccessReturnURL *string `json:"unlink_success_return_url"`
+
+	ServiceID  *string `json:"-"`
```
Comment (arlolanuzo) [inline pmsv2/params.go:47]: let's remove traces of `ServiceID`

File: pmsv2/qrcode/params.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+	QRString  string     `json:"qr_string"`
+	ExpiresAt *time.Time `json:"expires_at,omitempty"`
+}
+
+type CreateMethod struct {
+	Amount            float64               `json:"amount"`
+	Currency          constant.CurrencyEnum `json:"currency"`
+	ChannelCode       ChannelCode           `json:"channel_code"`
+	ChannelProperties ChannelProperties     `json:"channel_properties"`
+	QRCodeID          string                `json:"qr_code_id"`
```
Comment (arlolanuzo) [inline pmsv2/qrcode/params.go:33]: can remove

**Review by arlolanuzo — COMMENTED** [review]
sorry for another batch of changes 🙏

File: pmsv2/directdebit/params.go (hunk 8 above)
Comment (ligertsicat) [reply pmsv2/directdebit/params.go:7]: just to clarify, the ones in grey / Under Development should be removed, right?

File: pmsv2/client.go (hunk 5 above)
Comment (ligertsicat) [reply pmsv2/client.go]: removed

File: pmsv2/constant/params.go (hunk 7 above)
Comment (ligertsicat) [reply pmsv2/constant/params.go]: it even said internal use only haha

**Comment by cassiewang** [comment]
Please add `PaymentMethod` to list of sub packages in readme https://github.com/xendit/xendit-go/blob/master/README.md#sub-packages-documentations

**Comment by ligertsicat** [comment]
> Please add `PaymentMethod` to list of sub packages in readme https://github.com/xendit/xendit-go/blob/master/README.md#sub-packages-documentations

Added

File: pmsv2/directdebit/params.go (hunk 8 above)
Comment (ligertsicat) [reply pmsv2/directdebit/params.go:7]: oh alright. I've readded them

File: pmsv2/card/params.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+const (
+	CustomerUnscheduledCOF CardOnFileType = "CUSTOMER_UNSCHEDULED"
+	MerchantUnscheduledCOF CardOnFileType = "MERCHANT_UNSCHEDULED"
+	RecurringCOF           CardOnFileType = "RECURRING"
+)
+
+type CreateMethod struct {
+	Currency          *constant.CurrencyEnum `json:"currency"`
+	ChannelProperties *CardChannelProperties `json:"channel_properties"`
+	Token             *Token                 `json:"token"`
```
Comment (cassiewang) [inline pmsv2/card/params.go:18]: remove `Token`

File: pmsv2/directdebit/params.go (hunk 8 above)
Comment (arlolanuzo) [reply pmsv2/directdebit/params.go:7]: I think BCA_ONEKLIK, MTB and BDO are not yet supported

File: pmsv2/card/params.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+}
+type BillingDetails struct {
+	FirstName         *string `json:"first_name"`
+	AddressLine1      *string `json:"address_line_1,omitempty"`
+	AddressLine2      *string `json:"address_line_2,omitempty"`
+	AddressCity       *string `json:"address_city,omitempty"`
+	AddressCountry    *string `json:"address_country,omitempty"`
+	AddressPostalCode *string `json:"address_postal_code,omitempty"`
+	AddressState      *string `json:"address_state,omitempty"`
+}
```
Comment (cassiewang) [inline pmsv2/card/params.go:75]: these are wrong

File: pmsv2/client.go [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+func (c *Client) CreatePaymentMethodWithContext(ctx context.Context, data *CreatePaymentMethodParams) (*xendit.PaymentMethodResponse, *xendit.Error) {
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.XCallbackURL != "" {
+		header.Add("x-callback-url", data.XCallbackURL)
+	}
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
```
Comment (cassiewang) [inline pmsv2/client.go:32]: remove this part

File: pmsv2/directdebit/params.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+	BCAKlikpay ChannelCode = "BCA_KLIKPAY"
+)
+
+const (
+	BankAccountType  Type = "BANK_ACCOUNT"
+	DebitCardType    Type = "DEBIT_CARD"
+	BankRedirectType Type = "BANK_REDIRECT"
+)
+
+type CreateMethod struct {
```
Comment (arlolanuzo) [inline pmsv2/directdebit/params.go:37]: should only be channel_code and channel_properties

File: pmsv2/client.go (hunk 14 above)
Comment (cassiewang) [reply pmsv2/client.go:32]: and anywhere else where `business-id` and `x-callback-url` are added to headers

File: pmsv2/client.go [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+	return c.AuthPaymentMethodWithContext(context.Background(), data)
+}
+
+func (c *Client) AuthPaymentMethodWithContext(ctx context.Context, data *ValidateOTPRequest) (*xendit.PaymentMethodResponse, *xendit.Error) {
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
```
Comment (cassiewang) [inline pmsv2/client.go:60]: remove

File: pmsv2/ewallet/params.go [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+	DANA      ChannelCode = "DANA"
+	LinkAja   ChannelCode = "LINKAJA"
+	Shopeepay ChannelCode = "SHOPEEPAY"
+	Sakuku    ChannelCode = "SAKUKU"
+	Nexcash   ChannelCode = "NEXCASH"
+	Astrapay  ChannelCode = "ASTRAPAY"
+	Jeniuspay ChannelCode = "JENIUSPAY"
+)
+
+type CreateMethod struct {
```
Comment (arlolanuzo) [inline pmsv2/ewallet/params.go:21]: should only be channel_code and channel_properties also

File: pmsv2/client.go [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+	return c.ExpirePaymentMethodWithContext(context.Background(), data)
+}
+
+func (c *Client) ExpirePaymentMethodWithContext(ctx context.Context, data *ExpireRequest) (*xendit.PaymentMethodResponse, *xendit.Error) {
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
```
Comment (cassiewang) [inline pmsv2/client.go:88]: remove

File: pmsv2/client.go [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+func (c *Client) RetrievePaymentMethodWithContext(ctx context.Context, data *RetrievePaymentMethodRequest) (*xendit.PaymentMethodResponse, *xendit.Error) {
+
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
```
Comment (cassiewang) [inline pmsv2/client.go:117]: remove

File: pmsv2/client.go [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+	return c.UpdatePaymentMethodWithContext(context.Background(), data)
+}
+
+func (c *Client) UpdatePaymentMethodWithContext(ctx context.Context, data *UpdateRequest) (*xendit.PaymentMethodResponse, *xendit.Error) {
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
```
Comment (cassiewang) [inline pmsv2/client.go:192]: remove

File: pmsv2/client.go [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+func (c *Client) RetrievePaymentsWithContext(ctx context.Context, data *RetrievePaymentsRequest) (*xendit.PaymentMethodResponse, *xendit.Error) {
+
+	response := &xendit.PaymentMethodResponse{}
+	header := http.Header{}
+
+	if data.MerchantID != "" {
+		header.Add("business-id", data.MerchantID)
+	}
```
Comment (cassiewang) [inline pmsv2/client.go:221]: remove

File: pmsv2/params.go [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+	Type        constant.PaymentMethodTypeEnum `json:"type"`
+	Country     constant.CountryEnum           `json:"country"`
+	CustomerID  *string                        `json:"customer_id"`
+	ReferenceID string                         `json:"reference_id,omitempty"`
+	Reusability constant.ReusabilityEnum       `json:"reusability"`
+	Description *string                        `json:"description"`
+	Metadata    map[string]interface{}         `json:"metadata"`
+
+	XCallbackURL string `json:"-"`
+	MerchantID   string `json:"-"`
```
Comment (cassiewang) [inline pmsv2/params.go:25]: remove

File: pmsv2/params.go [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+	OverTheCounter *overthecounter.CreateMethod `json:"over_the_counter"`
+	VirtualAccount *virtualaccount.CreateMethod `json:"virtual_account"`
+	QRCode         *qrcode.CreateMethod         `json:"qr_code"`
+}
+
+type ValidateOTPRequest struct {
+	ID   string `param:"id"`
+	Code string `json:"auth_code"`
+
+	MerchantID string `json:"-"`
```
Comment (cassiewang) [inline pmsv2/params.go:39]: remove

Comment (cassiewang) [reply pmsv2/params.go:39]: ditto for other `MerchantID string `json:"-"`` 😅

File: pmsv2/card/params.go (hunk 12 above)
Comment (arlolanuzo) [reply pmsv2/card/params.go:18]: you think we should include `card_information` here?

**Comment by cassiewang** [comment]
Also have to support the headers listed in API ref: https://developers.xendit.co/api-reference/#create-payment-method

for POST requests:
- Idempotency-key
- for-user-id

for GET and PATCH requests:
  - for-user-id
  
e.g: https://github.com/xendit/xendit-go/blob/c12673fe6f1a9b9233598466f16c48f927e4bc28/directdebit/directdebitpayment/client.go#L86-L90

File: pmsv2/card/params.go (hunk 12 above)
Comment (cassiewang) [reply pmsv2/card/params.go:18]: told ligert what to add

File: payment_method.go [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+
+type EwalletMethod struct {
+	ChannelCode       ewallet.ChannelCode       `json:"channel_code"`
+	ChannelProperties ewallet.ChannelProperties `json:"channel_properties"`
+	Account           ewallet.Account           `json:"account"`
+}
+
+type DirectDebitMethod struct {
+	LinkedAccountTokenID string                        `json:"-"`
+	LinkedAccountID      string                        `json:"-"`
```
Comment (cassiewang) [inline payment_method.go:61]: remove

File: payment_method.go [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+	Currency          constant.CurrencyEnum    `json:"currency"`
+	ChannelCode       qrcode.ChannelCode       `json:"channel_code"`
+	ChannelProperties qrcode.ChannelProperties `json:"channel_properties"`
+}
+
+type PaymentMethodResponse struct {
+	ID                 string                           `json:"id"`
+	Type               constant.PaymentMethodTypeEnum   `json:"type"`
+	Country            constant.CountryEnum             `json:"country"`
+	MerchantID         string                           `json:"business_id"`
```
Comment (cassiewang) [inline payment_method.go:96]: let's call this `BusinessID` so it matches the json and it's also called `BusinessID` elsewhere in the repo

eg https://github.com/xendit/xendit-go/blob/c12673fe6f1a9b9233598466f16c48f927e4bc28/card.go#L15

File: pmsv2/directdebit/params.go (hunk 15 above)
Comment (ligertsicat) [reply pmsv2/directdebit/params.go:34]: updated to be only channel_code and channel_properties

File: pmsv2/ewallet/params.go (hunk 17 above)
Comment (ligertsicat) [reply pmsv2/ewallet/params.go:21]: updated to be only channel_code and channel_properties

File: pmsv2/params.go (hunk 23 above)
Comment (ligertsicat) [reply pmsv2/params.go]: done

File: pmsv2/card/params.go (hunk 13 above)
Comment (ligertsicat) [reply pmsv2/card/params.go]: updated based on the proxy payment method object

File: pmsv2/directdebit/params.go (hunk 8 above)
Comment (ligertsicat) [reply pmsv2/directdebit/params.go:7]: removed. 

Based on the reference, I don't see BCA_KLIKPAY either. Is that supported yet?

Comment (arlolanuzo) [reply pmsv2/directdebit/params.go:7]: right let's remove

Comment (ligertsicat) [reply pmsv2/directdebit/params.go:7]: removed

File: pmsv2/params.go [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+	ID   string `param:"id"`
+	Code string `json:"auth_code"`
+
+	ForUserID      string `json:"-"`
+	IdempotencyKey string `json:"-"`
+}
+
+type ExpireRequest struct {
+	ID                     string  `json:"-"`
+	UnlinkSuccessReturnURL *string `json:"unlink_success_return_url"`
```
Comment (cassiewang) [inline pmsv2/params.go:46]: please remove this for now, not live yet

Comment (ligertsicat) [reply pmsv2/params.go]: done

**Review by cassiewang — APPROVED** [review]

**Pull request merged**